Move over to Zola. (#10)
linebender/linebender.github.io: content/_index.md

+++
title = "Homepage"
description = "Homepage for the Linebender organization"
+++

Welcome to the homepage for the [Linebender organization](https://github.com/linebender). We are a friendly group of people who share an interest in 2D graphics and user interface design, with everything that entails. We hang out on [the Xi Zulip instance][xi.zulip] and always welcome new people.

# Linebender projects

Below is a list of the main linebender projects, and a short description of each project's purpose, as of 2023-04-22.

## Crates (actively developed)

These crates are under active development.

 - [`xilem`][xilem] - An experimental Rust architecture for reactive UI.

>    This crate is a spiritual successor to Druid, aiming to keep what is good while

Reviewer: ```suggestion
Xilem is a new UI toolkit with a medium-grained reactive architecture strongly inspired by SwiftUI. It is an evolution of Druid, aiming to keep what is good while improving on Druid's shortcomings.
```

Reviewer: I might have botched the diff here, I intended it to replace lines 18-21.

 - [`peniko`][peniko] - A library for non-geometric drawing primitives.

   This crate provides a set of shared types for concepts that are important for drawing/stroking paths, but excluding the path geometry itself (which can be found in [`kurbo`][kurbo]). It includes types for brush styles (including gradient) and color.

The following crates are not part of the `xilem`/`vello` stack, but are under active development.

 - [`velato`][velato] - A crate that renders lottie animations to vello.
 - [`norad`][norad] - A crate for reading, writing, and manipulating [Unified Font Object] files, a common font-design format.

## Crates (passively maintained)

 - [`druid`][druid] - A GUI framework for Rust

>    `druid` is the original Linebender UI framework, and was the main focus of development until recently. There are many good things about `druid`, but the data model it uses (primarily `Lens`es) gets complicated in Rust, and doesn't compose as well as it would in a [GC'd][garbage collection] functional programming language. However, a lot of the things that work will will be copied straight over to `xilem`, and today `druid` is a solid choice for a new UI app projected to have low to medium complexity. The [#druid-help channel on Zulip](https://xi.zulipchat.com/#narrow/stream/255910-druid-help) is still monitored and you will likely get good answers to any quesitons you have.

Reviewer: ```suggestion
   `druid` is the original Linebender UI framework, and was the main focus of development until recently. There are many good things about `druid`, but the data model it uses (primarily `Lens`es) gets complicated in Rust, and doesn't compose as well as it would in a [GC'd][garbage collection] functional programming language. However, a lot of the things that work will will be copied straight over to `xilem`, and today `druid` is a solid choice for a new UI app projected to have low to medium complexity. The [#druid-help channel on Zulip](https://xi.zulipchat.com/#narrow/stream/255910-druid-help) is still monitored and you will likely get good answers to any questions you have.
```

 - [`piet`][piet] - An abstraction layer over platform 2D rendering facilities.

>    The purpose of the `piet` library is to provide abstraction over the libraries used for 2D rendering on any particular platform ([Direct2D] on Windows, [Core Graphics] on Mac, and [cairo]/[pango] on Linux). The goal is for `piet` to be superseeded by the `vello` crate, which will provide state-of-the-art 2D rendering on any platform supported by [WebGPU].

Reviewer: ```suggestion
   The purpose of the `piet` library is to provide abstraction over the libraries used for 2D rendering on any particular platform ([Direct2D] on Windows, [Core Graphics] on Mac, and [cairo]/[pango] on Linux). The goal is for `piet` to be superseded by the `vello` crate, which will provide state-of-the-art 2D rendering on any platform supported by [WebGPU].
```

 - [`runebender`][runebender] - An experimental font editor.

   This was the motivating application for `druid` development. It is not currently seeing active development, but this may be revisited in the future.

 - [`skribo`][skribo] - A library for text layout.

   This library aims to perform the conversion from text and font attributes to glyph runs, similar to [harfbuzz]. It's not recommended to use this library as better alternatives exist, for example [`parley`] or [`cosmic-text`].

## Non-code repos

 - [`2d.graphics`][2d.graphics] - A work-in-progress book explaining the current state of the art of different aspects of 2D graphics.

   Currently the book mostly contains annotated reference lists, although some sections have written content. Pull requests welcome! The goal is to cover topics like color, shapes/curves, 2D GPU rendering, text, etc.

> 

Reviewer: Should this repo also be listed here?
